pkg/project/auth/cache.go

```diff
@@ -0,0 +1,482 @@
+	skip := false
+	if skip {
+		return
+	}
+
+	// by default, we update our current caches and do an incremental change
+	userSubjectRecordStore := ac.userSubjectRecordStore
+	groupSubjectRecordStore := ac.groupSubjectRecordStore
+	reviewRecordStore := ac.reviewRecordStore
+
+	// if there was a global change that forced complete invalidation, we rebuild our cache and do a fast swap at end
+	invalidateCache := ac.invalidateCache()
+	if invalidateCache {
+		userSubjectRecordStore = cache.NewStore(subjectRecordKeyFunc)
+		groupSubjectRecordStore = cache.NewStore(subjectRecordKeyFunc)
+		reviewRecordStore = cache.NewStore(reviewRecordKeyFunc)
+	}
+
+	// iterate over caches and synchronize our three caches
+	namespaceSet := ac.synchronizeNamespaces(userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+	ac.synchronizePolicies(userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+	ac.synchronizePolicyBindings(userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+	purgeDeletedNamespaces(namespaceSet, userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore)
+
+	// if we did a full rebuild, now we swap the fully rebuilt cache
+	if invalidateCache {
+		ac.userSubjectRecordStore = userSubjectRecordStore
+		ac.groupSubjectRecordStore = groupSubjectRecordStore
+		ac.reviewRecordStore = reviewRecordStore
+	}
+
+}
+
+// syncRequest takes a reviewRequest and determines if it should update the caches supplied, it is not thread-safe
+func (ac *AuthorizationCache) syncRequest(request *reviewRequest, userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store, reviewRecordStore cache.Store) error {
+
+	lastKnownValue, err := lastKnown(reviewRecordStore, request.namespace)
+	if err != nil {
+		return err
+	}
+
+	if skipReview(request, lastKnownValue) {
+		return nil
+	}
+
+	namespace := request.namespace
+	review, err := ac.reviewer.Review(namespace)
+	if err != nil {
+		return err
+	}
+
+	usersToRemove := util.NewStringSet()
+	groupsToRemove := util.NewStringSet()
+	if lastKnownValue != nil {
+		usersToRemove.Insert(lastKnownValue.users...)
+		usersToRemove.Delete(review.Users()...)
+		groupsToRemove.Insert(lastKnownValue.groups...)
+		groupsToRemove.Delete(review.Groups()...)
+	}
+
+	deleteSubjectsToNamespace(userSubjectRecordStore, usersToRemove.List(), namespace)
+	deleteSubjectsToNamespace(groupSubjectRecordStore, groupsToRemove.List(), namespace)
+	addSubjectsToNamespace(userSubjectRecordStore, review.Users(), namespace)
+	addSubjectsToNamespace(groupSubjectRecordStore, review.Groups(), namespace)
+	cacheReviewRecord(request, lastKnownValue, review, reviewRecordStore)
+	return nil
+}
+
+// List returns the set of namespace names the user has access to view
+func (ac *AuthorizationCache) List(userInfo user.Info) (*kapi.NamespaceList, error) {
+	keys := util.StringSet{}
+	user := userInfo.GetUID()
```

Comment:
Will fix that up and try to get a unit test with different values to make sure it doesn't happen again.